### ISSUE TITLE
Which HTML UA code bases need to support Web Media APIs?

### ISSUE THREAD

**Issue by mavgit** [w3c-cg/webmediaapi#27]
Breaking out this discussion from issue #6 into it's own issue.

I believe specs included in Web Media APIs spec must be supported by the main HTML User Agent (UA) code bases that are used in consumer products.

For the 2017 version of the spec, specs must be supported by the following code bases:
- Blink/Chromium
- Edge
- Gecko
- WebKit

**Comment by poolec** [comment]
Where we're talking about a browser feature that requires extensive code within the core of the browser engine or has low level platform dependencies, then it may be reasonable to omit it from the 2017 spec if the main code bases cannot support it.

However, where we're talking about specific APIs that don't go so deep, then I think we are missing an opportunity and creating a legacy for the future if we omit helpful APIs that are very useful for content providers, are widely supported (but missing from one browser), and can be straightforwardly be added (perhaps even as a javascript library).

It would be a missed opportunity if the 2017 spec just documented the results of CanIUse.com and did nothing to move things forward.

**Comment by poolec** [comment]
Discussion 15/12/2016:
Views expressed:
- WAVE should stay within the API sets supported by the four main code bases
- APIs should be considered on merit, balancing the cost of implementing vs. the value of the API to application developers
- Polyfills are often possible.  Should they be the responsibility of the application developers or the CE device manufacturers?

**Comment by boyofgreen** [comment]
Polyfills should be a responsibility of the developer.  Device manufactures may recommend particular ones based on their implementation (they will know what is missing), but in the end, the developer should make the decision to use a polyfill or work around the feature.

**Comment by mavgit** [comment]
Call for consensus:
1. For the 2017 version of the spec, specs must be supported by the following code bases:

- Blink/Chromium
- Edge
- Gecko
- WebKit

2. Polyfills are the responsibility of the developer. 

3. The spec should Recommend APIs that are desired, but not fully supported across all four code bases.

**Comment by jpiesing** [comment]
Is this based on the current versions of these code bases?
Browsers integrated in devices will be based on older versions of webkit / chromium.

**Comment by mavgit** [comment]
@jpiesing 

> Is this based on the current versions of these code bases?
Browsers integrated in devices will be based on older versions of webkit / chromium.

This is another issue which is worth discussing.

I suggest we close this issue with consensus on the three points in the call for consensus, because it marks important progress to have agreement on the four code bases, on polyfills and on recommendations.

Please open a new issue regarding versions of code bases. You can assume this issue as a base to build on.

(FWIW: One question I'd hope you could address is: What do we need to say anything about versions of code bases beyond that the code base complies with our spec requirements?)

**Comment by jpiesing** [comment]
@mavgit If we are making decisions based on a feature being supported in 4 code bases then if it was only recently added to one of the code bases used in devices (i.e. webkit or blink/chromium) then the ports of those code bases to devices for 2017 products won't have that feature.

**Comment by mavgit** [comment]
@jpiesing OK, then we can discuss the code base version issue here before closing this issue.

I propose to add a fourth item to the call for consensus, which I believe reflects previous discussions:

4. WAVE won't have any explicit version number requirement nor any version date requirement for any client code bases. Any code base that meets our spec requirements and passes our tests is WAVE HATF compliant, regardless of version number or how recent the code base is.

Two side notes, not part of the proposed updated call for consensus:

Side note 1: This does mean that for each code base there will be an implicit version number that is the earliest release with full WAVE HATF spec support. We could (and in my opinion should) publish the earliest compliant version numbers for our four target code bases as helpful, informative information for manufacturers choosing a code base, but we shouldn't require or test for any of these version numbers, because code bases can be branched without changing the version number, making any version number test unreliable. This should be discussed as a new issue, not here.

Side note 2: I also think we should include recommendations that manufacturers should use recent versions and update versions in the field where possible, to benefit from security fixes and bug fixes (as discussed in issue #30), but I don't think we can have any requirements or tests in this area. This should be discussed under issue #30, not here.

**Comment by mavgit** [comment]
Call for consensus, v2:

1. For the 2017 version of the spec, Required specs must be supported by all of the following code bases:
   - Blink/Chromium
   - Edge
   - Gecko
   - WebKit

2. Polyfills are the responsibility of the developer.

3. The spec should Recommend APIs that are desired, but not fully supported across all four code bases.

4. WAVE won't have any explicit version number requirement nor any version date requirement for any client code bases. Any code base that meets our spec requirements and passes our tests is WAVE HATF compliant, regardless of version number or how recent the code base is.

**Comment by jpiesing** [comment]
@mavgit I suggest the following modification to the call for consensus, change;
> For the 2017 version of the spec, Required specs must be supported by all of the following code bases:

to

> For the 2017 version of the spec, Required specs must be supported by all of the following code bases as of March 2016:

March 2016 is a guess and it would be nice if a browser porting / integration company like Espial or Opera TV could comment on it. A feature added to webkit / chromium-blink in (for example) September 2016 will likely be too late for the final delivery from our supplier to us for final integration for 2017 products.

**Comment by jeremypoulter** [comment]
Another option you could also look at is use a project like
https://github.com/cefsharp/CefSharp as one of your references, this adds
similar integration delays (two in fact) as you might see in 'real'
products and will take in to account the fact that some releases may be
more or less effort to integrate in to downstream code bases.

**Comment by davemevans** [comment]
> it would be nice if a browser porting / integration company like Espial or Opera TV could comment on it

We haven't had a response from any relevant companies, but it would be good to come to some consensus on this.

We are already making exceptions for known unsupported features, and it's difficult to see how we could be absolutely sure that all features required were available in all supported code bases at an arbitrary date.

**Comment by mavgit** [comment]
In reviewing this issue, I believe the problem is that the spec is currently claiming to define when products ship. The opening sentence is:
> This specification defines the Web APIs that should be included in device implementations to support media web apps in 2017.

If we were to maintain this claim, the spec would need to describe code availability at some unknown previous time, as proposed by @jpiesing https://github.com/w3c/webmediaapi/issues/27#issuecomment-284991701

However, as discussed above, we don't have any data about the time lag from code availability to shipping products. Certainly, that time lag will vary by company. It's also not the practice of W3C specs to be defining product shipping dates.

So, I propose to remove from the spec any language that discusses when products ship. This spec should define a set of APIs that are supported across the four code bases at the time the spec is published. This spec won't discuss when products will ship based on this spec.

I propose we change the opening sentence to:
> This specification defines the Web APIs to support media web apps that are supported across all four of the most widely used user agent code bases at the time of publication.

**Comment by mavgit** [comment]
Call for consensus, v3 (only item 5 is new):

1. For the 2017 version of the spec, Required specs must be supported by all of the following code bases:

- Blink/Chromium
- Edge
- Gecko
- WebKit

2. Polyfills are the responsibility of the developer.

3. The spec should Recommend APIs that are desired, but not fully supported across all four code bases.

4. WAVE won't have any explicit version number requirement nor any version date requirement for any client code bases. Any code base that meets our spec requirements and passes our tests is WAVE HATF compliant, regardless of version number or how recent the code base is.

5. Change the opening sentence to:
> This specification defines the Web APIs to support media web apps that are supported across all four of the most widely used user agent code bases at the time of publication.

**Comment by davemevans** [comment]
Consensus not reached on item 5 from v3.
Next step: Create new issue to discuss item 5.

Consensus reached on call 05/04/17 for items 1 - 4 from v2.
Next steps: create and merge a PR steps 1-4 as detailed in https://github.com/w3c/webmediaapi/issues/27#issuecomment-284910213, and close issue.

**Comment by davemevans** [comment]
Points 1 and 3 are already covered in the document:

> Required APIs are meant to be available in 2016 in a usable form in the four major web user agent code bases. This is to enable the APIs to be delivered in consumer products in 2017

> Recommended APIs represent desired functionality that is not yet implemented widely enough to be Required.

I have proposed additional text to close points 2 and 4 in #65.

If #65 is merged, we will close this issue and open a new issue for point 5.

**Comment by davemevans** [comment]
Consensus reached to merge #65 on telco 03/05/17 - merged as c1effb8.

I will open a new issue to address point 5 and close this once that action is complete.

**Comment by davemevans** [comment]
Raised https://github.com/w3c/webmediaapi/issues/70 to discuss point 5.
